From a71f3f93611e83ff1490099aafa4eed46b7e9504 Mon Sep 17 00:00:00 2001
From: Victor Bo <bvoid@yandex.ru>
Date: Wed, 5 Feb 2020 14:33:45 +0200
Subject: [PATCH 10/11] rework info about security patch date

  * based on
  https://github.com/LineageOS/android_packages_apps_Settings/commit/2fe6f351f98cb9456359182bcf902546b3dc638b#diff-fc3d9d607f2fdcc77bb9e90ba41e7389

	modified:   res/values/strings.xml
	modified:   res/xml/firmware_version.xml
	modified:   src/com/android/settings/deviceinfo/firmwareversion/SecurityPatchLevelPreferenceController.java
	new file:   src/com/android/settings/deviceinfo/firmwareversion/SecurityPatchVendorPreferenceController.java
---
 res/values/strings.xml                             |  2 +
 res/xml/firmware_version.xml                       |  9 ++-
 .../SecurityPatchLevelPreferenceController.java    | 60 +++++++++---------
 .../SecurityPatchVendorPreferenceController.java   | 73 ++++++++++++++++++++++
 4 files changed, 108 insertions(+), 36 deletions(-)
 create mode 100644 src/com/android/settings/deviceinfo/firmwareversion/SecurityPatchVendorPreferenceController.java

diff --git a/res/values/strings.xml b/res/values/strings.xml
index 839c892ec3..cc028708b8 100644
--- a/res/values/strings.xml
+++ b/res/values/strings.xml
@@ -3089,6 +3089,8 @@
     <string name="firmware_version">Android version</string>
     <!-- About phone screen, status item label  [CHAR LIMIT=60] -->
     <string name="security_patch">Android security patch level</string>
+    <!-- About phone screen, status item label  [CHAR LIMIT=60] -->
+    <string name="security_patch_vendor">Vendor security patch level</string>
     <!-- About phone screen, status item label  [CHAR LIMIT=40] -->
     <string name="model_info">Model</string>
     <!-- About phone screen, status item summary  [CHAR LIMIT=40] -->
diff --git a/res/xml/firmware_version.xml b/res/xml/firmware_version.xml
index 20dd034c80..17bf2b4709 100644
--- a/res/xml/firmware_version.xml
+++ b/res/xml/firmware_version.xml
@@ -37,13 +37,12 @@
         settings:enableCopying="true"
         settings:controller="com.android.settings.deviceinfo.firmwareversion.SecurityPatchLevelPreferenceController"/>
 
-    <!-- Mainline module version -->
+    <!-- Vendor security patch -->
     <Preference
-        android:key="module_version"
-        android:title="@string/module_version"
-        android:summary="@string/summary_placeholder"
+        android:key="security_key_vendor"
+        android:title="@string/security_patch_vendor"
         settings:enableCopying="true"
-        settings:controller="com.android.settings.deviceinfo.firmwareversion.MainlineModuleVersionPreferenceController"/>
+        settings:controller="com.android.settings.deviceinfo.firmwareversion.SecurityPatchVendorPreferenceController"/>
 
     <!-- Baseband -->
     <Preference
diff --git a/src/com/android/settings/deviceinfo/firmwareversion/SecurityPatchLevelPreferenceController.java b/src/com/android/settings/deviceinfo/firmwareversion/SecurityPatchLevelPreferenceController.java
index 1df78a8081..029e850a51 100644
--- a/src/com/android/settings/deviceinfo/firmwareversion/SecurityPatchLevelPreferenceController.java
+++ b/src/com/android/settings/deviceinfo/firmwareversion/SecurityPatchLevelPreferenceController.java
@@ -17,59 +17,57 @@
 package com.android.settings.deviceinfo.firmwareversion;
 
 import android.content.Context;
-import android.content.Intent;
-import android.content.pm.PackageManager;
-import android.net.Uri;
-import android.text.TextUtils;
-import android.util.Log;
-
-import androidx.preference.Preference;
+import android.os.SystemProperties;
+import android.text.format.DateFormat;
 
+import com.android.settings.R;
 import com.android.settings.core.BasePreferenceController;
-import com.android.settingslib.DeviceInfoUtils;
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Locale;
 
 public class SecurityPatchLevelPreferenceController extends BasePreferenceController {
 
     private static final String TAG = "SecurityPatchCtrl";
-    private static final Uri INTENT_URI_DATA = Uri.parse(
-            "https://source.android.com/security/bulletin/");
 
-    private final PackageManager mPackageManager;
-    private final String mCurrentPatch;
+    private static final String KEY_AOSP_SECURITY_PATCH =
+            "ro.build.version.security_patch";
+
+    private static final String KEY_AOSP_SECURITY_PATCH_COPY =
+            "build.version.security_patch";
 
     public SecurityPatchLevelPreferenceController(Context context, String key) {
         super(context, key);
-        mPackageManager = mContext.getPackageManager();
-        mCurrentPatch = DeviceInfoUtils.getSecurityPatch();
     }
 
     @Override
     public int getAvailabilityStatus() {
-        return !TextUtils.isEmpty(mCurrentPatch)
-                ? AVAILABLE : CONDITIONALLY_UNAVAILABLE;
+        return AVAILABLE;
     }
 
     @Override
     public CharSequence getSummary() {
-        return mCurrentPatch;
-    }
+        String patchLevel = SystemProperties.get(KEY_AOSP_SECURITY_PATCH_COPY);
 
-    @Override
-    public boolean handlePreferenceTreeClick(Preference preference) {
-        if (!TextUtils.equals(preference.getKey(), getPreferenceKey())) {
-            return false;
+        if (patchLevel.isEmpty()) {
+            patchLevel = SystemProperties.get(KEY_AOSP_SECURITY_PATCH);
         }
 
-        final Intent intent = new Intent();
-        intent.setAction(Intent.ACTION_VIEW);
-        intent.setData(INTENT_URI_DATA);
-        if (mPackageManager.queryIntentActivities(intent, 0).isEmpty()) {
-            // Don't send out the intent to stop crash
-            Log.w(TAG, "queryIntentActivities() returns empty");
-            return true;
+        if (!patchLevel.isEmpty()) {
+            try {
+                SimpleDateFormat template = new SimpleDateFormat("yyyy-MM-dd");
+                Date patchLevelDate = template.parse(patchLevel);
+                String format = DateFormat.getBestDateTimePattern(Locale.getDefault(), "dMMMMyyyy");
+                patchLevel = DateFormat.format(format, patchLevelDate).toString();
+            } catch (ParseException e) {
+                // parsing failed, use raw string
+            }
+        } else {
+            patchLevel = mContext.getString(R.string.unknown);
         }
 
-        mContext.startActivity(intent);
-        return true;
+        return patchLevel;
     }
 }
diff --git a/src/com/android/settings/deviceinfo/firmwareversion/SecurityPatchVendorPreferenceController.java b/src/com/android/settings/deviceinfo/firmwareversion/SecurityPatchVendorPreferenceController.java
new file mode 100644
index 0000000000..7ed48b1aa6
--- /dev/null
+++ b/src/com/android/settings/deviceinfo/firmwareversion/SecurityPatchVendorPreferenceController.java
@@ -0,0 +1,73 @@
+/*
+ * Copyright (C) 2019 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.settings.deviceinfo.firmwareversion;
+
+import android.content.Context;
+import android.os.SystemProperties;
+import android.text.format.DateFormat;
+
+import com.android.settings.R;
+import com.android.settings.core.BasePreferenceController;
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Locale;
+
+public class SecurityPatchVendorPreferenceController extends BasePreferenceController {
+
+    private static final String TAG = "SecurityPatchVendorCtrl";
+
+    private static final String KEY_VENDOR_SECURITY_PATCH =
+            "ro.vendor.build.security_patch";
+
+    private static final String KEY_XXX_SECURITY_PATCH =
+            "ro.keymaster.xxx.security_patch";
+
+    public SecurityPatchVendorPreferenceController(Context context, String key) {
+        super(context, key);
+    }
+
+    @Override
+    public int getAvailabilityStatus() {
+        return AVAILABLE;
+    }
+
+    @Override
+    public CharSequence getSummary() {
+        String patchLevel = SystemProperties.get(KEY_XXX_SECURITY_PATCH);
+
+        if (patchLevel.isEmpty()) {
+            patchLevel = SystemProperties.get(KEY_VENDOR_SECURITY_PATCH);
+        }
+
+        if (!patchLevel.isEmpty()) {
+            try {
+                SimpleDateFormat template = new SimpleDateFormat("yyyy-MM-dd");
+                Date patchLevelDate = template.parse(patchLevel);
+                String format = DateFormat.getBestDateTimePattern(Locale.getDefault(), "dMMMMyyyy");
+                patchLevel = DateFormat.format(format, patchLevelDate).toString();
+            } catch (ParseException e) {
+                // parsing failed, use raw string
+            }
+        } else {
+            patchLevel = mContext.getString(R.string.unknown);
+        }
+
+        return patchLevel;
+    }
+}
-- 
2.11.0

